opencog/atomspace/AtomTable.h

```diff
@@ -232,12 +242,26 @@ class AtomTable
                         bool subclass = false) const
     {
         std::lock_guard<std::recursive_mutex> lck(_mtx);
+
+        // We're going to use this one handle so we don't construct/destruct
+        // googles of temporaries in the process of doing the copies.
+        Handle temp_handle;
         std::for_each(typeIndex.begin(type, subclass),
                       typeIndex.end(),
-             [&](Handle h)->void {
-                  if (not isDefined(h)) return;
-                  (func)(h);
-             });
+            [&](UUID uuid)->void {
+                if (UNDEFINED_UUID == uuid) return;
+                
+                // Since the function could and probably will resolve the
+                // AtomPtr, we need to reset the handle since we are going to
+                // bypass the normal shared_ptr semantics here to avoid
+                // a copy per handle which is much more expensive than
+                // resetting and setting a new UUID. So we call the safe
+                // version which takes a bit longer.
+                temp_handle.set_uuid(uuid);
```

Comment:
I'm going to revise my prior statement: "No, copying is most certainly much less than resolve." In thinking about it, the "most certainly" is just plain dumb of me.

The fact, is, it is possible that it is cheaper to do an unordered_set hash lookup with unsigned in keys than to do several shared_ptr operations and construction / destruction calls along with a memory copy of the members.

I'll test this and see. I've been surprised when my assumptions didn't hold enough to prefer testing in cases that actually matter.